File: flutter/plugin/lib/serializer/serializer.dart
import 'package:xplpc/message/decoded_request.dart';
import 'package:xplpc/message/param.dart';

abstract class Serializer {
  DecodedRequest? decodeRequest(String data);
  String encodeFunctionReturnValue(dynamic data);
  T? decodeFunctionReturnValue<T>(String data);
  String encodeRequest(String functionName, [List<Param>? params]);
}

File: kotlin/shared/src/com/xplpc/serializer/Serializer.kt
package com.xplpc.serializer

import com.google.gson.reflect.TypeToken
import com.xplpc.message.DecodedRequest
import com.xplpc.message.Param

interface Serializer {
    fun encodeRequest(functionName: String, vararg params: Param): String
    fun <T> decodeFunctionReturnValue(data: String, type: TypeToken<T>): T?
    fun decodeRequest(data: String): DecodedRequest?
    fun encodeFunctionReturnValue(data: Any?): String
}
